Fix milestone 1 of #921: allow jumping to a particular state in preview mode

oparry opened the pull request:
Stay on the same state of an exploration when switching between edit and preview modes. (Milestone 1 of #921).

    When switching to preview, the URL is constructed using the current state name.
    The URL is reconstructed when moving to the next state in preview mode (correct answer submitted).
    When a preview URL is detected, the state name is extracted and the preview tab is refreshed accordingly.

Explorations with parameters are not properly handled yet, so 3. always begins the preview in the initial state. N.B. - the URL is still constructed using the active state, so it is wrong until returning to edit mode or moving on to the next state.

seanlip commented:
NB: this supersedes #1708.

@oparry, please let me know (by writing a top-level comment in this thread) once you've addressed all the individual comments on the other PR, and I'll be happy to take a second review pass!

seanlip commented:
Hi @oparry -- ping?

seanlip commented on core/templates/dev/head/editor/EditorServices.js:
```diff
@@ -46,8 +46,8 @@ oppia.factory('explorationData', [
       // Returns a promise that supplies the data for the current exploration.
-      getData: function() {
-        if (explorationData.data) {
+      getData: function(ignoreCache=false) {
```
Default params don't seem to be well-supported: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters.

Let's not use them.

oparry replied:
Removed.

seanlip commented on core/templates/dev/head/editor/ExplorationPreview.js:
```diff
@@ -18,19 +18,29 @@
     $scope.isExplorationPopulated = false;
-    explorationData.getData().then(function() {
+    explorationData.getData(ignoreCache = true).then(function() {
+      var stateName = editorContextService.getActiveStateName();
```
initStateName is probably a better variable name here.

oparry replied:
Changed.

seanlip commented on core/templates/dev/head/editor/ExplorationPreview.js:
```diff
@@ -18,19 +18,29 @@
       explorationGadgetsService, oppiaPlayerService) {
     $scope.isExplorationPopulated = false;
-    explorationData.getData().then(function() {
+    explorationData.getData(ignoreCache = true).then(function() {
```
Would you mind helping me understand, in more detail, why ignoreCache is needed? I confess that I don't really get it...

I.e., when you say "the state wouldn't update properly", what exactly do you mean? The main change we're making here is which initial state the preview is starting from. I don't see why this has any bearing on the exploration data being retrieved.

oparry replied:
It isn't needed.
I managed to convince myself that it was before the last commit... I think I must have added the parameter at the same time as fixing a different bug.  Sorry for the confusion!

seanlip commented on core/templates/dev/head/editor/RouterServices.js:
```diff
@@ -70,26 +71,7 @@ oppia.factory('routerService', [
         _tabs.active = FEEDBACK_TAB;
       } else if (newPath.indexOf('/gui/') !== -1) {
```
I like what you did above (=== 0). Maybe do it here too?

oparry replied:
Done.

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -315,7 +315,9 @@ oppia.factory('oppiaPlayerService', [
             }
           }
-
+          if (_editorPreviewMode) {
```
Actually, I'm not sure if we need to do this at all? There are two ways to think about the URL here:

  (a) it represents the current state. If that's the case, though, then it needs to be updated if, e.g., the user clicks the back/forward buttons on the progress dots.

  (b) it represents the state that the preview starts from. Then we don't have to do anything here.

I lean towards the second, both for simplicity, and also because many pages sort of work that way (e.g. you search in the gallery, and then you scroll down; more explorations load, but the URL doesn't change). What do you think? We can still do something to keep track of the current state (maybe in editorContextService) and load the correct state in the gui view based on that.

oparry replied:
ok. I'm fine with b), so I've removed this now, but I definitely think we should keep track of the state in preview mode so it can be returned to in the gui view.  Would I need to call editorContextService.setActiveStateName() in the same place I was changing the path, and also for the case where the back/forward buttons are clicked?

seanlip replied:
Good question. I think -- yes, probably. That seems to make sense.

seanlip commented on core/templates/dev/head/editor/RouterServices.js:
```diff
@@ -137,34 +155,34 @@ oppia.factory('routerService', [
-        _savePendingChanges();
-        $location.path('/preview');
+      navigateToPreviewTab: function(stateName) {
+        if (_tabs.active === PREVIEW_TAB) {
```
Hm. If this is the case -- maybe just do nothing? We'll have a reset button later.

Otherwise, you're bringing DOM references into a JS service, which is weird.

seanlip replied:
Ah, no -- for the main tab, that code is there to handle changes between states in the exploration, since that's a common action on that tab. For the preview tab, this isn't really the case, so it's fine to remove it from here, but I would leave the main tab stuff as it is for now. Thanks for asking!

oparry replied:
ok, thanks, I've removed that.  I also removed the stateName parameter, since it is never used.

seanlip commented:
Hi @oparry -- thanks! The main outstanding problem with this PR is the ignoreCache stuff -- it's causing the frontend tests to fail in Travis-CI, for one thing. I don't actually understand why it's needed, so if you could elaborate a bit more on what's going on, that would be great.

The other comments are minor, and I think you did a great job with the refactoring -- it looks very nice!

Also, copying the comments from #1708 here -- we should do the following changes before this PR can be merged into develop:

- handle explorations with parameters correctly (with the "set parameters" modal) -- this is milestone 2
- add a little box at the top left, or a warning message at the top, saying "Starting preview from STATE_NAME" if the starting state is not the initial state for the exploration. (This needs a little UI design thinking to make it look nice, but I do think that the user needs to be made aware of this fact.)

Thanks!

seanlip commented on core/templates/dev/head/editor/RouterServices.js:
```diff
@@ -118,6 +124,18 @@ oppia.factory('routerService', [
     };
 
+    var _actuallyNavigate = function(pathType, newStateName) {
+      if (pathType != 'preview' && pathType != 'gui') {
```
Probably nicer to use the existing constants: MAIN_TAB, PREVIEW_TAB

oparry replied:
MAIN_TAB is currently set as 'main' rather than 'gui'.  Could it be defined as 'gui' instead?

seanlip replied:
Oh, sorry, you're right. My bad.

Let's define two new constants: SLUG_GUI and SLUG_PREVIEW, and use those instead. A slug is the 'end part' of a URL. The existing constants are, I believe, already used for other things, and you don't want to just change them because this may break other parts of the app.

seanlip commented on core/templates/dev/head/editor/RouterServices.js:
```diff
@@ -137,34 +155,26 @@ oppia.factory('routerService', [
         if (_tabs.active === MAIN_TAB) {
           $('.oppia-editor-cards-container').fadeOut(function() {
-            _actuallyNavigate(stateName);
+            _actuallyNavigate('gui', stateName);
```
Use MAIN_TAB, PREVIEW_TAB (here and below)?

seanlip commented:
Hi @oparry, this looks good! I had only a couple minor comments.

Just to confirm: could you please also do the other two changes in this PR? FYI we are planning to do a release around end-April/beginning-May, so hopefully we can get the PR merged in by then.

Thanks!

oparry commented:
Hi @seanlip, 

- I’ve added a box to warn the user when the preview is starting from a state other than the initial one.

<img width="914" alt="preview_jump-to-state_warning_example" src="https://cloud.githubusercontent.com/assets/9349009/14792615/fa806790-0b12-11e6-9335-b04908ae0e38.png">

Are there any particular css classes/styles I should be using?  For now I just grabbed class="popover sharp-borders fade bottom in" and style="top: 56px; left: 16px; display: block;" from elsewhere in the code.

- At the moment, the warning is displayed if the active state is the initial preview state, but not the first state of the exploration.  For that to work, I need to keep the active state updated in preview mode.  The following approach seems to work; could you tell me if it is reasonable?

1.  Add editorContextService to ProgressDotsDirective.js and update the active state name when changeActiveDot() is called.
2. Do a $rootScope.$broadcast('updateActiveState’) in oppiaPlayerService.submitAnswer(), watch for the broadcast in ExplorationPreview.js, and update the active state there.

Finally, would you mind giving me a hint as to how to bring up the set parameters modal?

Thanks!

seanlip commented:
Hi @oparry,

1. Hm, it could look nicer, but I'm not sure which classes/styles to recommend. It might be a good idea to define a new style for this box, and tweak its CSS -- it seems to be quite a separate thing from the other components on the site. I would add a bit more whitespace around the text; it feels a bit too near the edges at the moment.

2. I think you might as well keep the warning throughout -- it seems like a useful reference. Otherwise, it seems a bit weird to have it reappearing when the user navigates back to the initial preview state. I think this means that you can set a $scope variable in the exploration preview controller storing the message to show at the top left, and just keep that there. What do you think?

3. To bring up the modal, you'll probably want to use $modal.open, and add a template for the contents of the modal. See [this PR](https://github.com/oppia/oppia/pull/1755/files) by @anthkris for a recent example. (Does that answer your question?)

seanlip commented:
Hi @oparry -- how is this going?

oparry commented:
Hi @seanlip,
Really sorry for the delay with this - I haven't had a chance to work on it over the last week.  I have some time tomorrow and over the weekend, so I'll try and update on Saturday or Sunday.

seanlip commented:
OK, thanks @oparry!

oparry commented:
Hi @seanlip,
I've completed the things we discussed above, except for creating the modal to enter undefined parameters.  I think I now understand the basics of how to make a modal template and create an instance with $modal.open, but I'm not sure if I'm adding the code in the right place.  Could you tell me if the following sounds reasonable? So far I've:
1. Created a template for the modal and put it in a new file in core/templates/dev/head/components/ .
2. Added an {% include ... statement to exploration_preview.html to include the modal html .
3. Added a function to the ExplorationPreview controller that creates an instance of the modal.

Presumably then I need a trigger in the controller to call the modal-creating function when a required parameter is undefined?  I can't see how I could check which parameters are required in the current state though...
Would I also need to validate the inputs before setting the new parameter values?

seanlip commented:
Hi @oparry -- yes, sounds good, except that, for 2, I'd put the {% include... %} in exploration_editor.html instead, with all the other components, since that's the "root" page.

For validation, I can help out a bit here. There's a method in EditorServices.js called _verifyParameters() that should do what you want (you can pass in the relevant initNodeIds), but it currently returns data in a form that's not very useful for your purposes. I'll refactor it to fix this, and let you know once I've done this. For now, pretend you have a black-box method to which you can pass a list of initial state names and that will give you a list of parameters that need to be set for the exploration to be valid from that point on.

All parameters are currently raw strings, so no, you don't need to validate the inputs. However, we are likely to move to typed parameters in the future, so please add a TODO at the relevant point to add input validation.

Thanks!

seanlip commented:
Hi @oparry, just a quick heads-up -- it might be a good idea to try and get this in soon, if you're able? I'm planning to start reorganizing the files in /editor to separate things into different directories, so if that gets in before this one, it's probably going to lead to some largish merge conflicts, since all the preview-related files will have been moved elsewhere. Thanks!

oparry commented:
Hi @seanlip, thanks for the heads up.  I'm think I'm almost done with this, but I'm struggling with the details of how to modify the parameters and load the initial preview state.  At the moment I
1. Create the modal instance with an input for each parameter that needs to be set,
2. Call LearnerParamsService.setValue() for each one,
3. Use a callback to refresh the state with oppiaPlayerService.populateExploration().

But this will only work if I also call oppiaPlayerService.populateExploration() before opening the modal, otherwise I get "Cannot set unknown parameter:"  In addition, although it seems to set the parameters themselves correctly, any values displayed in the new state that depend on those parameters don't get updated.  Is there a better way to go about this?

seanlip commented:
Hm, sounds like this needs some debugging. Could you push your changes somewhere (maybe a separate branch) and provide steps to reproduce? I'll try to take a look at it later.

Thanks!

oparry commented:
Changes pushed to https://github.com/oparry/oppia/commit/2f169df609e0e6c0d524299fe02aaa8fe1eb6288.
To reproduce:
- Load "The Lazy Magician" exploration
- Open editor and go to "Magician%20Guesses" state 
- Hit preview
- Enter values for "lowest" and "highest" and click ok.
Debugging text suggests lowest and highest are correctly updated, but {{guess}} is 0 rather than {{floor((lowest+highest)/2)}}.

seanlip commented:
OK, I took a look! The main piece of advice I have is that it's really important to understand what's going on, rather than just trying a bunch of things and hoping that something works. (E.g. I noticed you were experimenting with calling loadPreviewState() twice, which doesn't really seem like a good idea conceptually.) Often, understanding what's going on requires some digging and investigation, but this digging can't really be skipped; an _ad hoc_ solution gives no reassurance that it will work in all cases. (E.g. the commit you linked to fails when you try to preview starting from the first card.)

So, I did some digging, and here's what currently happens in develop (it's a good idea to verify this for yourself so you understand how the flow works, too):

- ExplorationPreview calls populateExploration(), which provides PlayerServices with the data needed to construct the exploration object, and then turns on the isExplorationPopulated flag, which causes the conversation skin directive to load.
- When the conversation skin directive loads, it calls playerServices.init(). This does the following:
  - It zeroes out the LearnerParamsService.
  - It calls _loadInitialState(), which gets the list of all parameters used in the exploration and sets them using the parameter changes that are to be applied at the start of the exploration. (These are defined in the Settings tab.)

So, with this in mind, what do we do? It's a bit hard for ExplorationPreview to "reach into" the conversation skin directive and meddle with the init() call. You kind of want ExplorationPreview to set everything up nicely, and then conversation_skin should be able to just run on autopilot from there.

This suggests redefining populateExploration() and init() in PlayerServices. In particular, populateExploration() seems to be a function that's only used in editor preview mode, so it looks like a good place for all the editor-preview-mode stuff. So, what I would suggest is extending the definition of populateExploration() to do the following:
- load the current version of the exploration in the editor (which it already does)
- zero out the LearnerParamsService
- get the list of all params used in the exploration and set them using the param changes that are to be applied at the start of the exploration -- but do _not_ apply the parameter changes related to the initial state. The latter is common functionality between the learner and the editor preview modes, so it might as well go in init().

Then, the scope of init() and _loadInitialState() would be reduced accordingly to contain only the stuff that is common to both types of initializations (i.e. from the learner view, and from the editor preview mode).

Does this make sense? I can make a PR to develop that does a refactor along these lines, but I figured I'd explain the thought process so that you have a sense of how one comes up with a solution to this kind of problem.

seanlip commented:
Added aforementioned PR in #1838.

oparry commented:
Thanks a lot for explaining all of that, I hadn't appreciated that refactoring would need to be done elsewhere in the code.

I completely agree that making ad hoc changes with no clear understanding of what's going on is an unreliable and inefficient way of doing things! Unfortunately, I don't have the knowledge of the codebase, or of Angular, to do much else at the moment.  I think I also misunderstood the concept of a starter project and assumed that the changes would be mostly cosmetic.  In my defence, that commit was really only intended to show you what I had so far, not to be a working solution!

With a new language, I would usually spend time finding a decent IDE, learning syntax and best practice in a more systematic way, and working out how to debug and follow code execution.  Since I can't commit the time to do those things at the moment, maybe I should stick to simpler frontend stuff, or python projects, where I have some previous experience?  Anyway, thanks for taking the time to help me understand the flow of the code, I really appreciate it.

seanlip commented:
Hi @oparry, no worries! In general, it's easy to just suggest that someone not make _ad hoc_ changes but not give much other guidance, so I thought it might be useful to explain concretely how one might work through an issue like this in a systematic way, without having to understand too much of the codebase in order to do so. Please don't take this as an indication that something's going badly wrong; it was just meant to be knowledge that you can apply in the future.

As for next steps, I've merged #1838 into develop, so when you get a chance could you please merge that into this branch and make the appropriate residual changes? I think this PR is getting pretty close to the finish line; the main thing is to test that the intended behavior works correctly and as expected in all cases.

And yep, sorry again that this project ended up a bit harder than expected! I just wanted to say, though, that I think you're handling it really well, and it's functionality that we're excited to ship soon. Also, FYI, we've just had a bunch of new starter projects come in, so the next one is likely to be simpler :-)

seanlip commented:
Hi @oparry, also, a quick heads-up -- per #1828, we have changed the editor/ folder to be exploration_editor/ instead, for parity with collection_editor/ -- and split up the editor tabs into individual folders. You might want to do a merge in order to update your local branch. Thanks!

oparry commented:
Hi @seanlip, 
I'm looking into why these two e2e tests fail, but haven't been able to figure it out yet.  Any advice would be much appreciated!

seanlip commented:
Hi @oparry! I can't tell why they're going wrong just by looking, but here's some pointers:

- The "unexpected alert open" probably just happens because the previous test failed. So I suggest focusing only on the other one.
- For the other test, the best thing to do is to set up a demo server and just manually follow the steps in the failing test in core/tests/protractor/editorAndPlayer.js. See if you get any console errors, or unexpected results -- that should give you a clue. The advantage of doing it manually is that you can look around and zero in on what's happening when you reproduce the error manually.

oparry commented:
Great, thanks for the pointers.  Could you explain how to set up a demo server and follow through the test?  Apologies if this is on the wiki somewhere, I couldn't see it.

seanlip commented:
Basically, run `bash scripts/start.sh`, and read through the test file. E.g. if it says editor.setStateName('card1'), then modify the state name to 'card1', etc.

It looks like the error occurs around the preview tab, so you might want to also try creating a simple exploration and viewing the preview tab starting from the various states, and making sure everything works correctly. If that's the case, then yeah, the best thing to do is to just follow along through the e2e test in order to better understand the sequence of actions that leads to the error.

seanlip commented:
(Note: the test file is in core/tests/protractor/editorAndPlayer.js. Have a look at the test logs in Travis to pinpoint the line of code in that file where the error occurs.)

oparry commented:
Thanks, I see what's going on now.  The test expects the preview to always start from the first card, rather than the state being edited, which is no longer the case after these commits.  When the preview begins, it expects the content to match that entered for the first card.  I think simply switching to the first state before previewing will solve the problem, but it's not easy to tell on my laptop because a lot of the e2e tests fail for other reasons.  Should I just commit that change and see if the Travis build succeeds?

seanlip commented:
Yes, that's fine! You might perhaps also consider adding a couple of lines to check that the correct state is shown when the preview starts from a different card. 

Feel free to make any additional commits you need and use Travis as a check for them.

oparry commented:
Hi @seanlip, I'm unable to reproduce this failure in the MISC E2E tests.  It passed in the previous build (#2018) and I can't see how the last commit could have affected it... Is it possible to force it to try the build again without making another commit?

souravbadami commented:
@oparry -- I've restarted the build.

oparry commented:
@souravbadami Thanks!

seanlip commented:
Tests look good now (thanks @souravbadami!). @oparry, is this ready for a review?

oparry commented:
Yep, all ready, thanks!

seanlip commented on core/templates/dev/head/exploration_editor/RouterServices.js:
```diff
@@ -98,6 +83,30 @@ oppia.factory('routerService', [
       }
     });
 
+    var _doNavigation = function(path, pathType) {
```
Maybe call this _doNavigationWithState, so that devs don't think it's a general purpose navigation function to be used for /settings, /history, etc.

oparry replied:
Done.

seanlip commented on core/templates/dev/head/exploration_editor/RouterServices.js:
```diff
@@ -137,34 +158,26 @@ oppia.factory('routerService', [
       navigateToPreviewTab: function() {
-        _savePendingChanges();
-        $location.path('/preview');
+        if (_tabs.active !== PREVIEW_TAB) {
```
Seems to have dropped _savePendingChanges(). Is this intentional? If you have a field open in the editor and then you move to the preview tab, does it get saved? (It should...)

oparry replied:
Oops, no, it wasn't intentional, I'll restore that.  The changes are saved anyway, presumably because of the $rootScope.$broadcast('externalSave') at line 56.  Should that really be a call to _savePendingChanges() too?  Or isn't the try/catch needed in that case?

seanlip replied:
Oh, interesting. I'm not sure. How about, for now, just restoring it anyway, but adding a comment above it:

  TODO(oparry): Determine whether this is necessary, since there's already a $rootScope.$broadcast('externalSave') in line 56

(note that the name next to the TODO refers to who is reporting it, not who should do it). Also, please file an issue on GitHub about this so we don't forget to look into it.

Thanks!

oparry replied:
ok, although I was thinking the other way around - since _savePendingChanges() is called by each of the tab navigation functions, is the broadcast at line 56 necessary?  The comment in _savePendingChanges() also suggests $rootScope.$broadcast('externalSave') on its own isn't guaranteed to work.

seanlip replied:
Ah. OK -- maybe put a comment there instead, then. I don't know offhand; it needs more investigation I think.

seanlip commented on core/templates/dev/head/exploration_editor/preview_tab/PreviewTab.js:
```diff
@@ -42,17 +123,21 @@ oppia.controller('PreviewTab', [
       }, 200);
+    };
+
+    $scope.$on('updateActiveState', function(evt, stateName) {
```
Could you explain again why this is here / why it is needed? It doesn't seem to tie into anything in particular so I'm wondering if all the updateActiveState event-passing is still necessary.

oparry replied:
This is intended to make sure the active state is kept up-to-date when navigating the exploration in preview mode, either by clicking through to the next card, or by clicking on the progress dots.  That way, when clicking back to the editor, the state doesn't change.  This wasn't listed as one of the milestones in #921, but it was mentioned as desirable behaviour in the discussion.  Personally, I think it is intuitive to always stay in the same state when switching between the editor and preview.

seanlip replied:
Oh, I see. In that case, can you rename this event to "updateActiveStateIfInEditor", and add a comment above this line explaining what you just mentioned? That should help avoid confusion.

oparry replied:
Done.

seanlip commented on core/templates/dev/head/exploration_editor/preview_tab/preview_tab.html:
```diff
@@ -14,3 +17,22 @@
+
+<style>
+  .oppia-preview-state-warning {
+    position: absolute;
```
nit: alphabetize these -- so, background-color, then border, then border-radius, etc. Makes it easier to find the value for a particular property.

oparry replied:
Done.

seanlip commented:
Hi @oparry -- this is awesome! I just played with it and it works beautifully.

I have only four small comments. Can you address them, and then we can get this merged? It's an important feature, so I expect that we'll also be calling it out in the next release email.

Thanks!

seanlip commented:
Hi @oparry, this is really really nice. Thanks so much! Merging :-)

oparry commented:
Great! Thanks for all your help with it.
